src/dctest/core.cljs

```diff
@@ -177,87 +179,119 @@ Options:
-                                    error (or (:error results)
-                                              (first (keep #(get-expect-error step-context %) (:expect step))))]
-                              (merge results
-                                     (when error {:error error}))))
-              {:keys [interval retries]} (:repeat step)
-              results (retry/retry-times run-attempt
-                                         retries
-                                         {:delay-ms interval
-                                          :check-fn #(not (:error %))})
-
-              outcome (if (:error results) :fail :pass)
-              results (merge results
-                             {:outcome outcome}
-                             (select-keys step [:name]))
-              context (update-in context [:state :failed] #(or % (failure? results)))]
+(defn get-expect-error [context expect]
+  (if (string? expect)
+    (when-not (expr/read-eval context expect)
+      {:message (str "Expectation failure: " expect)
+       :debug (expr/explain-refs context expect)})
+    (first (keep #(get-expect-error context %) expect))))
+
+(defn execute-step-attempts [{:keys [context results]}]
+  (P/let [{:keys [expect repeat]} results
+          {:keys [interval retries]} repeat
+          run-attempt (fn []
+                        (P/let [exec-results (run-exec context results)
+                                context (assoc context :step exec-results)
+                                error (or (:error exec-results)
+                                          (get-expect-error context expect))
+                                results (if error
+                                          (assoc results :outcome :fail :error error)
+                                          results)]
+                          {:context context :results results}))]
+    (retry/retry-times run-attempt
+                       retries
+                       {:delay-ms interval
+                        :check-fn #(not (failure? (:results %)))})))
+
+(defn execute-step [orig-context step]
+  (P/-> {:context orig-context
+         :results (assoc step :outcome :pending)}
+
+        (while-> (comp pending? :results)
+
+                 (read-eval-key :if)
+                 (update :results (fn [step]
+                                    (if (:if step)
+                                      step
+                                      (assoc step :outcome :skip))))
+
+                 (interpolate-key :env)
+                 (as-> m
+                   (update-in m [:context :env] merge (get-in m [:results :env])))
+
+                 (interpolate-key :exec)
+                 (interpolate-key :run)
+                 (update-in [:results :index] #(or % 1))
+
+                 (execute-step-attempts)
+
+                 (assoc-in [:results :outcome] :pass))
+
+        (assoc :context orig-context)
```

Comment:
This was the other area I thrashed on a lot.  I agree that the ideal would be the context is just threaded straight through.  There are definitely cases were tests or steps _do_ want to modify the context for future tests/steps.  The first one is to update the `[:context :state :failed]`, yup (and this could also be rethought).  The second one that we haven't added yet is that jobs and steps in GHA provide outputs that future jobs and steps can refer to.  These outputs would go into the context in our case.

The thing we need to be careful of at the moment is that any env in the current suite/test/step is not leaked into future suites/tests/steps.  Perhaps this is the only thing we actually "revert" here.  I considered having a stack (a vector) of envs that gets pushed to and popped off, but I didn't actually implement it to see if it was as weird as I expected it to be.  I'll try that.

Reply:
Addressed!  context is now only passed down, never "mutated" and returned back "up"